File: app/recycle_service.go
package main

// 业务编排：工程/单位的删除、回收站恢复与彻底删除（协调 DB 记录与磁盘文件）。

import (
	"os"
	"path/filepath"

	"heritage-mgmt/internal/domain"
)

// recycleFolderToBin 将工程归档目录移入回收站(保留文件)；目录不存在则忽略。
func recycleFolderToBin(folder string) {
	if folder == "" {
		return
	}
	src := filepath.Join(projectsDir, folder)
	if _, err := os.Stat(src); err != nil {
		return
	}
	bin := filepath.Join(dataDir, "recycle")
	os.MkdirAll(bin, 0755)
	os.Rename(src, filepath.Join(bin, folder))
}

// RecycleProject 软删除工程并将其文件移入回收站。
func RecycleProject(proj *domain.Project) error {
	if err := SoftDeleteProject(proj.ID); err != nil {
		return err
	}
	recycleFolderToBin(proj.Folder)
	return nil
}

// RestoreProject 从回收站恢复工程：取消软删除标记并把文件搬回。
func RestoreProject(proj *domain.Project) {
	RestoreProjectRecord(proj.ID)
	recyclePath := filepath.Join(dataDir, "recycle", proj.Folder)
	if _, err := os.Stat(recyclePath); err == nil {
		os.Rename(recyclePath, filepath.Join(projectsDir, proj.Folder))
	}
}

// PurgeProject 彻底删除工程：删除 DB 记录与回收站文件。
func PurgeProject(proj *domain.Project) {
	PurgeProjectRecord(proj.ID)
	os.RemoveAll(filepath.Join(dataDir, "recycle", proj.Folder))
}

// DeleteUnitCascade 删除单位：其下工程文件移入回收站，再删除单位与工程/文档的 DB 记录。
// 返回受影响的工程数。
func DeleteUnitCascade(uid int64) int {
	pids := ProjectIDsByUnit(uid)
	for _, pid := range pids {
		if proj, _ := GetProject(pid); proj != nil {
			recycleFolderToBin(proj.Folder)
		}
	}
	DeleteUnitRecords(uid)
	return len(pids)
}

File: desktop/src/core/recycle/RecycleService.cpp
#include "RecycleService.h"

#include <QDir>
#include <QFile>
#include <QFileInfo>

namespace heritage {

RecycleService::RecycleService(ProjectRepo& projects, UnitRepo& units, LogRepo& logs,
                               const AppConfig& cfg)
    : projects_(projects), units_(units), logs_(logs), cfg_(cfg) {}

void RecycleService::recycleFolderToBin(const QString& folder) {
    if (folder.isEmpty())
        return;
    const QString src = cfg_.projectsDir + QStringLiteral("/") + folder;
    if (!QFileInfo::exists(src))
        return;
    const QString bin = cfg_.dataDir + QStringLiteral("/recycle");
    QDir().mkpath(bin);
    // 目标若已存在先移除（避免 rename 失败）
    const QString dst = bin + QStringLiteral("/") + folder;
    if (QFileInfo::exists(dst))
        QDir(dst).removeRecursively();
    QDir().rename(src, dst);
}

bool RecycleService::recycleProject(const Project& p) {
    if (!projects_.softDelete(p.id))
        return false;
    recycleFolderToBin(p.folder);
    logs_.insert(QStringLiteral("删除工程"), QStringLiteral("工程#%1 %2").arg(p.id).arg(p.name), p.folder);
    return true;
}

void RecycleService::restoreProject(const RecycledProject& r) {
    projects_.restore(r.id);
    const QString recyclePath = cfg_.dataDir + QStringLiteral("/recycle/") + r.folder;
    if (QFileInfo::exists(recyclePath))
        QDir().rename(recyclePath, cfg_.projectsDir + QStringLiteral("/") + r.folder);
    logs_.insert(QStringLiteral("恢复工程"), QStringLiteral("工程#%1 %2").arg(r.id).arg(r.name), r.folder);
}

void RecycleService::purgeProject(const RecycledProject& r) {
    projects_.purge(r.id);
    QDir(cfg_.dataDir + QStringLiteral("/recycle/") + r.folder).removeRecursively();
    logs_.insert(QStringLiteral("彻底删除"), QStringLiteral("工程#%1 %2").arg(r.id).arg(r.name), r.folder);
}

int RecycleService::deleteUnitCascade(qint64 unitId) {
    const QVector<qint64> pids = projects_.idsByUnit(unitId);
    for (qint64 pid : pids) {
        const auto proj = projects_.get(pid);
        if (proj)
            recycleFolderToBin(proj->folder);
    }
    units_.deleteRecords(unitId);
    logs_.insert(QStringLiteral("删除单位"), QStringLiteral("单位#%1").arg(unitId),
                 QStringLiteral("级联%1个工程").arg(pids.size()));
    return pids.size();
}

} // namespace heritage
